Boosting refers to an ensemble method in which several models are trained sequentially with each model learning from the errors of its predecessors. In this chapter, you'll be introduced to the two boosting methods of AdaBoost and Gradient Boosting.

## Adaboost

#### Define the AdaBoost classifier
In the following exercises you'll revisit the Indian Liver Patient dataset which was introduced in a previous chapter. Your task is to predict whether a patient suffers from a liver disease using 10 features including Albumin, age and gender. However, this time, you'll be training an AdaBoost ensemble to perform the classification task. In addition, given that this dataset is imbalanced, you'll be using the ROC AUC score as a metric instead of accuracy.

As a first step, you'll start by instantiating an AdaBoost classifier.

In [1]:
import pandas as pd

indian_liver = pd.read_csv('indian_liver_patient_preprocessed.csv')
indian_liver

Unnamed: 0   Age_std  Total_Bilirubin_std  Direct_Bilirubin_std  \
0             0  1.247403            -0.420320             -0.495414   
1             1  1.062306             1.218936              1.423518   
2             2  1.062306             0.640375              0.926017   
3             3  0.815511            -0.372106             -0.388807   
4             4  1.679294             0.093956              0.179766   
..          ...       ...                  ...                   ...   
574         578  0.938909            -0.452462             -0.495414   
575         579 -0.295067            -0.436391             -0.495414   
576         580  0.445318            -0.404249             -0.459878   
577         581 -0.850356            -0.323893             -0.353271   
578         582 -0.418465            -0.372106             -0.424343   

     Alkaline_Phosphotase_std  Alamine_Aminotransferase_std  \
0                   -0.428870                     -0.355832   
1                    1.675083                     -0.093573   
2                    0.816243                     -0.115428   
3                   -0.449416                     -0.366760   
4                   -0.395996                     -0.295731   
..                        ...                           ...   
574                  0.857336                     -0.333977   
575                 -0.794596                     -0.252021   
576                 -0.190532                     -0.180993   
577                 -0.441198                     -0.284804   
578                 -0.309701                     -0.328514   

     Aspartate_Aminotransferase_std  Total_Protiens_std  Albumin_std  \
0                         -0.319111            0.293722     0.203446   
1                         -0.035962            0.939655     0.077462   
2                         -0.146459            0.478274     0.203446   
3                         -0.312205            0.293722     0.329431   
4                         -0.177537            0.755102    -0.930414   
..                              ...                 ...          ...   
574                       -0.263863           -0.536763    -1.938290   
575                       -0.274222           -0.444487     0.077462   
576                       -0.212067           -0.075383     0.077462   
577                       -0.270769            0.293722     0.329431   
578                       -0.298393            0.755102     1.589276   

     Albumin_and_Globulin_Ratio_std  Is_male_std  Liver_disease  
0                         -0.147390            0              1  
1                         -0.648461            1              1  
2                         -0.178707            1              1  
3                          0.165780            1              1  
4                         -1.713237            1              1  
..                              ...          ...            ...  
574                       -1.807188            1              0  
575                        0.478949            1              1  
576                        0.165780            1              1  
577                        0.165780            1              1  
578                        1.731627            1              0  

[579 rows x 12 columns]

In [2]:
indian_liver.drop(columns = 'Unnamed: 0', inplace= True)
indian_liver

Age_std  Total_Bilirubin_std  Direct_Bilirubin_std  \
0    1.247403            -0.420320             -0.495414   
1    1.062306             1.218936              1.423518   
2    1.062306             0.640375              0.926017   
3    0.815511            -0.372106             -0.388807   
4    1.679294             0.093956              0.179766   
..        ...                  ...                   ...   
574  0.938909            -0.452462             -0.495414   
575 -0.295067            -0.436391             -0.495414   
576  0.445318            -0.404249             -0.459878   
577 -0.850356            -0.323893             -0.353271   
578 -0.418465            -0.372106             -0.424343   

     Alkaline_Phosphotase_std  Alamine_Aminotransferase_std  \
0                   -0.428870                     -0.355832   
1                    1.675083                     -0.093573   
2                    0.816243                     -0.115428   
3                   -0.449416                     -0.366760   
4                   -0.395996                     -0.295731   
..                        ...                           ...   
574                  0.857336                     -0.333977   
575                 -0.794596                     -0.252021   
576                 -0.190532                     -0.180993   
577                 -0.441198                     -0.284804   
578                 -0.309701                     -0.328514   

     Aspartate_Aminotransferase_std  Total_Protiens_std  Albumin_std  \
0                         -0.319111            0.293722     0.203446   
1                         -0.035962            0.939655     0.077462   
2                         -0.146459            0.478274     0.203446   
3                         -0.312205            0.293722     0.329431   
4                         -0.177537            0.755102    -0.930414   
..                              ...                 ...          ...   
574                       -0.263863           -0.536763    -1.938290   
575                       -0.274222           -0.444487     0.077462   
576                       -0.212067           -0.075383     0.077462   
577                       -0.270769            0.293722     0.329431   
578                       -0.298393            0.755102     1.589276   

     Albumin_and_Globulin_Ratio_std  Is_male_std  Liver_disease  
0                         -0.147390            0              1  
1                         -0.648461            1              1  
2                         -0.178707            1              1  
3                          0.165780            1              1  
4                         -1.713237            1              1  
..                              ...          ...            ...  
574                       -1.807188            1              0  
575                        0.478949            1              1  
576                        0.165780            1              1  
577                        0.165780            1              1  
578                        1.731627            1              0  

[579 rows x 11 columns]

In [3]:
# Import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier

# Import AdaBoostClassifier
from sklearn.ensemble import AdaBoostClassifier

# Instantiate dt
dtc = DecisionTreeClassifier(max_depth=2, random_state=1)

# Instantiate ada
ada = AdaBoostClassifier(estimator=dtc, n_estimators=180, random_state=1)

Well done! Next comes training ada and evaluating the probability of obtaining the positive class in the test set.

In [4]:
from sklearn.model_selection import train_test_split

X = indian_liver.iloc[:,:-1]
y =  indian_liver.iloc[:,-1]
SEED=1
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, stratify = y, random_state = SEED)

#### Train the AdaBoost classifier
Now that you've instantiated the AdaBoost classifier ada, it's time train it. You will also predict the probabilities of obtaining the positive class in the test set. This can be done as follows:

Once the classifier ada is trained, call the .predict_proba() method by passing X_test as a parameter and extract these probabilities by slicing all the values in the second column as follows:

ada.predict_proba(X_test)[:,1]
The Indian Liver dataset is processed for you and split into 80% train and 20% test. Feature matrices X_train and X_test, as well as the arrays of labels y_train and y_test are available in your workspace. In addition, we have also loaded the instantiated model ada from the previous exercise.

In [19]:
# Fit ada to the training set
ada.fit(X_train, y_train)

# Compute the probabilities of obtaining the positive class
y_pred_proba = ada.predict_proba(X_test)[:,1]

Great work! Next, you'll evaluate ada's ROC AUC score.

#### Evaluate the AdaBoost classifier
Now that you're done training ada and predicting the probabilities of obtaining the positive class in the test set, it's time to evaluate ada's ROC AUC score. Recall that the ROC AUC score of a binary classifier can be determined using the roc_auc_score() function from sklearn.metrics.

The arrays y_test and y_pred_proba that you computed in the previous exercise are available in your workspace.

In [6]:
# Import roc_auc_score
from sklearn.metrics import roc_auc_score

# Evaluate test-set roc_auc_score
ada_roc_auc = roc_auc_score(y_test, y_pred_proba)

# Print roc_auc_score
print('ROC AUC score: {:.2f}'.format(ada_roc_auc),"\n\n")

ROC AUC score: 0.70 




Not bad! This untuned AdaBoost classifier achieved a ROC AUC score of 0.70!

## Gradient Boosting (GB)

#### Define the GB regressor
You'll now revisit the Bike Sharing Demand dataset that was introduced in the previous chapter. Recall that your task is to predict the bike rental demand using historical weather data from the Capital Bikeshare program in Washington, D.C.. For this purpose, you'll be using a gradient boosting regressor.

As a first step, you'll start by instantiating a gradient boosting regressor which you will train in the next exercise.

In [7]:
import pandas as pd
bikes = pd.read_csv('bikes.TXT')
bikes

hr  holiday  workingday  temp   hum  windspeed  cnt  instant  mnth  yr  \
0      0        0           0  0.76  0.66     0.0000  149    13004     7   1   
1      1        0           0  0.74  0.70     0.1343   93    13005     7   1   
2      2        0           0  0.72  0.74     0.0896   90    13006     7   1   
3      3        0           0  0.72  0.84     0.1343   33    13007     7   1   
4      4        0           0  0.70  0.79     0.1940    4    13008     7   1   
...   ..      ...         ...   ...   ...        ...  ...      ...   ...  ..   
1483  19        0           1  0.80  0.49     0.1343  452    14487     8   1   
1484  20        0           1  0.80  0.49     0.1343  356    14488     8   1   
1485  21        0           1  0.76  0.58     0.1940  303    14489     8   1   
1486  22        0           1  0.76  0.58     0.1940  277    14490     8   1   
1487  23        0           1  0.74  0.62     0.1045  174    14491     8   1   

      Clear to partly cloudy  Light Precipitation  Misty  
0                          1                    0      0  
1                          1                    0      0  
2                          1                    0      0  
3                          1                    0      0  
4                          1                    0      0  
...                      ...                  ...    ...  
1483                       1                    0      0  
1484                       1                    0      0  
1485                       1                    0      0  
1486                       1                    0      0  
1487                       1                    0      0  

[1488 rows x 13 columns]

In [8]:
from sklearn.model_selection import train_test_split

X = bikes.iloc[:,0:6].join(bikes.iloc[:, 7:-1])
y = bikes['cnt']

SEED = 1
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = SEED)

In [9]:
# Import GradientBoostingRegressor
from sklearn.ensemble import GradientBoostingRegressor

# Instantiate gb
gbr = GradientBoostingRegressor(max_depth=4, 
            n_estimators=200,
            random_state=2)

Awesome! Time to train the regressor and predict test set labels.

#### Train the GB regressor
You'll now train the gradient boosting regressor gb that you instantiated in the previous exercise and predict test set labels.

The dataset is split into 80% train and 20% test. Feature matrices X_train and X_test, as well as the arrays y_train and y_test are available in your workspace. In addition, we have also loaded the model instance gb that you defined in the previous exercise.

In [10]:
# Fit gb to the training set
gbr.fit(X_train, y_train)

# Predict test set labels
y_pred = gbr.predict(X_test)

### It gives an error of ' DataConversionWarning: A column-vector y was passed when a 1d array was expected. 
### Please change the shape of y to (n_samples, ), for example using ravel(). y = column_or_1d(y, warn=True)'
### Because In scikit-learn, the target variable y for regression models should be a 1D array, 
### while in this case, it seems to be a column vector.(If we used "y = bikes['cnt']" like that, 
### it would be an 1D array and no problem. bikes[['cnt']] is a dataFrame)

Great work! Time to evaluate the test set RMSE!

#### Evaluate the GB regressor
Now that the test set predictions are available, you can use them to evaluate the test set Root Mean Squared Error (RMSE) of gb.

y_test and predictions y_pred are available in your workspace.

In [11]:
# Import mean_squared_error as MSE
from sklearn.metrics import mean_squared_error as MSE

# Compute MSE
mse_test = MSE(y_test, y_pred)

# Compute RMSE
rmse_test = mse_test ** (1/2)

# Print RMSE
print('Test set RMSE of gb: {:.3f}'.format(rmse_test))

Test set RMSE of gb: 43.145


Great work!

## Stochastic Gradient Boosting (SGB)
#### Regression with SGB
As in the exercises from the previous lesson, you'll be working with the Bike Sharing Demand dataset. In the following set of exercises, you'll solve this bike count regression problem using stochastic gradient boosting.

In [12]:
# Import GradientBoostingRegressor
from sklearn.ensemble import GradientBoostingRegressor

# Instantiate sgbr
sgbr = GradientBoostingRegressor(max_depth=4, 
            subsample=0.9,
            max_features=0.75,
            n_estimators=200,
            random_state=2)

#### Train the SGB regressor
In this exercise, you'll train the SGBR sgbr instantiated in the previous exercise and predict the test set labels.

The bike sharing demand dataset is already loaded processed for you; it is split into 80% train and 20% test. The feature matrices X_train and X_test, the arrays of labels y_train and y_test, and the model instance sgbr that you defined in the previous exercise are available in your workspace.

In [13]:
# Fit sgbr to the training set
sgbr.fit(X_train, y_train)

# Predict test set labels
y_pred = sgbr.predict(X_test)

#### Evaluate the SGB regressor
You have prepared the ground to determine the test set RMSE of sgbr which you shall evaluate in this exercise.

y_pred and y_test are available in your workspace.

In [14]:
# Import mean_squared_error as MSE
from sklearn.metrics import mean_squared_error as MSE

# Compute test set MSE
mse_test = MSE(y_test, y_pred)

# Compute test set RMSE
rmse_test = mse_test ** (1/2)

# Print rmse_test
print('Test set RMSE of sgbr: {:.3f}'.format(rmse_test))

Test set RMSE of sgbr: 44.628


The stochastic gradient boosting regressor achieves a higher test set RMSE than the gradient boosting regressor (which was 43.145)! Normally, it shoud have been lower than gbr